src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,987 @@ mod v1 {
+        Ok(snapshot_meta)
+    }
+
+    fn check_snapshot_meta(snapshot_meta: &SnapshotMeta) -> RaftStoreResult<Vec<CfName>> {
+        let mut res = Vec::with_capacity(SNAPSHOT_CFS.len());
+        for cf_file_meta in snapshot_meta.get_cf_files() {
+            let found = SNAPSHOT_CFS.iter().find(|&s| cf_file_meta.get_cf() == *s);
+            if found.is_none() {
+                return Err(box_err!("failed to decode invalid snapshot cf {}",
+                                    cf_file_meta.get_cf()));
+            }
+            let cf_name = found.unwrap();
+            if res.iter().any(|cf| cf_name == cf) {
+                return Err(box_err!("failed to decode duplicated snapshot cf {}", cf_name));
+            }
+            res.push(*cf_name);
+        }
+        if res.len() != SNAPSHOT_CFS.len() {
+            return Err(box_err!("invalid number of snapshot meta cf file: {}, expect: {}",
+                                res.len(),
+                                SNAPSHOT_CFS.len()));
+        }
+        Ok(res)
+    }
+
+    fn check_file_size_and_checksum(path: &PathBuf,
+                                    expect_size: u64,
+                                    expect_checksum: u32)
+                                    -> RaftStoreResult<()> {
+        let size = try!(get_file_size(path));
+        if size != expect_size {
+            return Err(box_err!("invalid size {} for snapshot cf file {}, expected {}",
+                                size,
+                                path.display(),
+                                expect_size));
+        }
+        let checksum = try!(calc_crc32(&path));
+        if checksum != expect_checksum {
+            return Err(box_err!("invalid checksum {} for snapshot cf file {}, expected {}",
+                                checksum,
+                                path.display(),
+                                expect_checksum));
+        }
+        Ok(())
+    }
+
+    #[derive(Default)]
+    struct CfFile {
+        pub cf: CfName,
+        pub path: PathBuf,
+        pub tmp_path: PathBuf,
+        pub sst_writer: Option<SstFileWriter>,
+        pub file: Option<File>,
+        pub kv_count: u64,
+        pub size: u64,
+        pub written_size: u64,
+        pub checksum: u32,
+        pub write_digest: Option<Digest>,
+    }
+
+    #[derive(Default)]
+    struct MetaFile {
+        pub meta: SnapshotMeta,
+        pub path: PathBuf,
+        pub file: Option<File>,
+
+        // for writing snapshot
+        pub tmp_path: PathBuf,
+    }
+
+    pub struct Snap {
+        display_path: String,
+        cf_files: Vec<CfFile>,
+        cf_index: usize,
+        meta_file: MetaFile,
+        size_track: Arc<RwLock<u64>>,
+    }
+
+    impl Snap {
+        fn new<T: Into<PathBuf>>(dir: T,
+                                 key: &SnapKey,
+                                 size_track: Arc<RwLock<u64>>,
+                                 cfs: &[CfName],
```

Comment:
When receiving a snapshot, the cf sequence is specified by the received snapshot meta. `cfs` is not `SNAPSHOT_CFS` then.